File: rust/crates/oxa-anthropic/tests/nonstream.rs
//! Nonstream behavior beyond the golden vectors: structural error
//! boundaries, loss details, the max_tokens default, the string shorthand,
//! and verbatim tool-input bytes (INV-1).

use oxa_anthropic::{
    Config, Request, Response, decode_request, decode_response, encode_request, encode_response,
};
use oxa_ir::{LossReason, StopReason};
use serde_json::Value;

fn wire_request(json: Value) -> Request {
    serde_json::from_value(json).expect("test wire request must deserialize")
}

fn wire_response(json: Value) -> Response {
    serde_json::from_value(json).expect("test wire response must deserialize")
}

fn minimal_request() -> Value {
    serde_json::json!({
        "model": "claude-sonnet-4-5",
        "max_tokens": 128,
        "messages": [{ "role": "user", "content": "Hello" }]
    })
}

fn loss_with<'a>(losses: &'a [oxa_ir::Loss], path: &str, field: &str) -> &'a oxa_ir::Loss {
    losses
        .iter()
        .find(|loss| loss.path == path && loss.field == field)
        .unwrap_or_else(|| panic!("loss {path}.{field} not found in {losses:?}"))
}

#[test]
fn max_tokens_must_be_positive() {
    let mut wire = minimal_request();
    wire["max_tokens"] = serde_json::json!(0);
    let err = decode_request(&wire_request(wire), &Config::default()).expect_err("zero max_tokens");
    assert!(err.to_string().contains("max_tokens is required"), "{err}");
}

#[test]
fn unknown_role_is_a_structural_error() {
    let mut wire = minimal_request();
    wire["messages"] = serde_json::json!([{ "role": "chief", "content": "Hello" }]);
    let err = decode_request(&wire_request(wire), &Config::default()).expect_err("unknown role");
    assert!(err.to_string().contains("unknown role \"chief\""), "{err}");
}

#[test]
fn request_without_messages_is_a_structural_error() {
    let mut wire = minimal_request();
    wire["messages"] = serde_json::json!([]);
    let err = decode_request(&wire_request(wire), &Config::default()).expect_err("no messages");
    assert!(err.to_string().contains("carries no messages"), "{err}");
}

#[test]
fn missing_stop_reason_is_a_structural_error() {
    let wire = wire_response(serde_json::json!({
        "id": "r", "type": "message", "role": "assistant", "model": "m",
        "content": [{ "type": "text", "text": "hi" }],
        "stop_reason": ""
    }));
    let err = decode_response(&wire, &Config::default()).expect_err("missing stop reason");
    assert!(err.to_string().contains("stop_reason is missing"), "{err}");
}

#[test]
fn unsupported_system_block_type_is_a_structural_error() {
    let mut wire = minimal_request();
    wire["system"] = serde_json::json!([{ "type": "image", "text": "nope" }]);
    let err = decode_request(&wire_request(wire), &Config::default()).expect_err("system type");
    assert!(
        err.to_string().contains("unsupported block type \"image\""),
        "{err}"
    );
}

#[test]
fn tool_use_without_id_name_or_object_input_are_errors() {
    let base = serde_json::json!({
        "model": "m",
        "max_tokens": 8,
        "messages": [
            { "role": "assistant", "content": [{ "type": "tool_use", "name": "f", "input": {} }] }
        ]
    });
    let err =
        decode_request(&wire_request(base.clone()), &Config::default()).expect_err("missing id");
    assert!(
        err.to_string()
            .contains("messages[0].content[0].id is required"),
        "{err}"
    );

    let mut no_name = base.clone();
    no_name["messages"][0]["content"][0] =
        serde_json::json!({ "type": "tool_use", "id": "t", "input": {} });
    let err = decode_request(&wire_request(no_name), &Config::default()).expect_err("missing name");
    assert!(err.to_string().contains(".name is required"), "{err}");

    let mut no_input = base;
    no_input["messages"][0]["content"][0] =
        serde_json::json!({ "type": "tool_use", "id": "t", "name": "f" });
    let err =
        decode_request(&wire_request(no_input), &Config::default()).expect_err("missing input");
    assert!(err.to_string().contains(".input is required"), "{err}");
}

#[test]
fn tool_choice_named_requires_a_name() {
    let mut wire = minimal_request();
    wire["tool_choice"] = serde_json::json!({ "type": "tool" });
    let err = decode_request(&wire_request(wire), &Config::default()).expect_err("decode name");
    assert!(
        err.to_string().contains("tool_choice.name is required"),
        "{err}"
    );

    let req = oxa_ir::Request {
        model: "m".to_string(),
        system: Vec::new(),
        messages: vec![oxa_ir::Message {
            role: oxa_ir::Role::User,
            content: vec![oxa_ir::Block::Text {
                text: "hi".to_string(),
            }],
        }],
        tools: None,
        tool_choice: Some(oxa_ir::ToolChoice {
            mode: oxa_ir::ToolChoiceMode::Tool,
            name: None,
        }),
        params: Some(oxa_ir::Params {
            temperature: None,
            top_p: None,
            max_tokens: Some(8),
            stop_sequences: None,
        }),
        metadata: None,
    };
    let err = encode_request(&req, &Config::default()).expect_err("encode name");
    assert!(
        err.to_string().contains("tool_choice.name is required"),
        "{err}"
    );
}

#[test]
fn unknown_stop_reason_maps_to_other_with_a_loss() {
    let wire = wire_response(serde_json::json!({
        "id": "r", "type": "message", "role": "assistant", "model": "m",
        "content": [{ "type": "text", "text": "hi" }],
        "stop_reason": "model_context_window_exceeded"
    }));
    let (resp, losses) = decode_response(&wire, &Config::default()).expect("decode");
    assert_eq!(resp.stop_reason, StopReason::Other);
    let loss = loss_with(&losses, "stop_reason", "stop_reason");
    assert_eq!(loss.reason, LossReason::UnmappedValue);
}

#[test]
fn response_missing_content_is_a_structural_error() {
    let wire = wire_response(serde_json::json!({
        "id": "r", "type": "message", "role": "assistant", "model": "m",
        "content": [{ "type": "tool_result", "tool_use_id": "t" }],
        "stop_reason": "end_turn"
    }));
    let err = decode_response(&wire, &Config::default()).expect_err("missing content");
    assert!(
        err.to_string().contains("content[0].content is missing"),
        "{err}"
    );
}

#[test]
fn missing_max_tokens_defaults_to_4096_with_a_degraded_loss() {
    let req = oxa_ir::Request {
        model: "m".to_string(),
        system: Vec::new(),
        messages: vec![oxa_ir::Message {
            role: oxa_ir::Role::User,
            content: vec![oxa_ir::Block::Text {
                text: "hi".to_string(),
            }],
        }],
        tools: None,
        tool_choice: None,
        params: None,
        metadata: None,
    };
    let (wire, losses) = encode_request(&req, &Config::default()).expect("encode");
    assert_eq!(wire.max_tokens, 4096);
    let loss = loss_with(&losses, "params", "max_tokens");
    assert_eq!(loss.reason, LossReason::Degraded);
    assert!(loss.detail.contains("defaulting to 4096"), "{loss:?}");
}

#[test]
fn single_text_message_renders_the_string_shorthand() {
    let req = oxa_ir::Request {
        model: "m".to_string(),
        system: Vec::new(),
        messages: vec![oxa_ir::Message {
            role: oxa_ir::Role::User,
            content: vec![oxa_ir::Block::Text {
                text: "hi".to_string(),
            }],
        }],
        tools: None,
        tool_choice: None,
        params: Some(oxa_ir::Params {
            temperature: None,
            top_p: None,
            max_tokens: Some(8),
            stop_sequences: None,
        }),
        metadata: None,
    };
    let (wire, _) = encode_request(&req, &Config::default()).expect("encode");
    assert_eq!(
        wire.messages[0].content,
        Some(oxa_anthropic::ContentValue::Text("hi".to_string()))
    );

    // With a system prompt the shorthand no longer applies.
    let mut with_system = req;
    with_system.system = vec![oxa_ir::SystemBlock::Text {
        text: "be brief".to_string(),
    }];
    let (wire, _) = encode_request(&with_system, &Config::default()).expect("encode");
    assert!(matches!(
        wire.messages[0].content,
        Some(oxa_anthropic::ContentValue::Blocks(_))
    ));
}

#[test]
fn encoding_ir_stop_other_is_a_structural_error() {
    let resp = oxa_ir::Response {
        id: "r".to_string(),
        model: "m".to_string(),
        content: Vec::new(),
        stop_reason: StopReason::Other,
        stop_sequence: None,
        usage: oxa_ir::Usage {
            input_tokens: 0,
            output_tokens: 0,
        },
    };
    let err = encode_response(&resp, &Config::default()).expect_err("stop other");
    assert!(err.to_string().contains("no Anthropic equivalent"), "{err}");
}

#[test]
fn unsupported_tool_result_content_is_dropped_with_a_loss() {
    let req = oxa_ir::Request {
        model: "m".to_string(),
        system: Vec::new(),
        messages: vec![oxa_ir::Message {
            role: oxa_ir::Role::User,
            content: vec![oxa_ir::Block::ToolResult {
                tool_use_id: "call_1".to_string(),
                content: vec![oxa_ir::Block::ToolUse {
                    id: "nested_1".to_string(),
                    name: "weather".to_string(),
                    input: r#"{"city":"Paris"}"#.to_string(),
                }],
                is_error: None,
            }],
        }],
        tools: None,
        tool_choice: None,
        params: Some(oxa_ir::Params {
            temperature: None,
            top_p: None,
            max_tokens: Some(8),
            stop_sequences: None,
        }),
        metadata: None,
    };

    let (wire, losses) = encode_request(&req, &Config::default()).expect("encode");
    let Some(oxa_anthropic::ContentValue::Blocks(blocks)) = &wire.messages[0].content else {
        panic!("tool_result must render as a block array");
    };
    let Some(oxa_anthropic::ContentValue::Blocks(content)) = &blocks[0].content else {
        panic!("tool_result content must render as a block array");
    };
    assert!(content.is_empty());
    let loss = loss_with(&losses, "messages[0].content[0].content[0]", "type");
    assert_eq!(loss.reason, LossReason::UnsupportedSemantic);
}

#[test]
fn block_content_deserializes_raw_tool_input_verbatim() {
    let content: oxa_anthropic::ContentValue =
        serde_json::from_str(r#"[{"type":"tool_use","id":"t1","name":"f","input":{"x":1e+01}}]"#)
            .expect("block content must deserialize");
    let oxa_anthropic::ContentValue::Blocks(blocks) = content else {
        panic!("array content must remain blocks");
    };
    assert_eq!(
        blocks[0]
            .input
            .as_deref()
            .map(serde_json::value::RawValue::get),
        Some(r#"{"x":1e+01}"#)
    );
}

#[test]
fn request_json_deserialization_canonicalizes_tool_input_in_generic_content() {
    // Anthropic request `messages[].content` follows the reference face's
    // generic content boundary. The intermediate value has no raw source
    // bytes, so tool input is serialized canonically before reaching the
    // typed block representation.
    let wire: Request = serde_json::from_str(
        r#"{
            "model":"claude-sonnet-4-5",
            "max_tokens":128,
            "messages":[{
                "role":"assistant",
                "content":[{
                    "type":"tool_use",
                    "id":"t1",
                    "name":"f",
                    "input":{"x":1e+01}
                }]
            }]
        }"#,
    )
    .expect("deserialize");
    let (req, _) = decode_request(&wire, &Config::default()).expect("decode");
    let oxa_ir::Block::ToolUse { input, .. } = &req.messages[0].content[0] else {
        panic!("expected tool_use block");
    };
    assert_eq!(input, r#"{"x":10.0}"#);

    let (encoded, _) = encode_request(&req, &Config::default()).expect("encode");
    let encoded = serde_json::to_string(&encoded).expect("serialize");
    assert!(
        encoded.contains(r#""input":{"x":10.0}"#),
        "input remains canonical in the wire JSON: {encoded}"
    );
}

File: ts/src/anthropic/messages/nonstream.ts
import { OxaError } from "../../error.js";
import {
  fromValue,
  integer,
  isJsonArray,
  isJsonNumber,
  jsonText,
  parseJson,
  sourceTextOf,
  stringifyJson,
  withSourceText,
  type JsonObject,
  type JsonText,
  type JsonValue,
} from "../../json/index.js";
import type { ConversionResult, Loss, LossReason } from "../../loss.js";
import { mapModel, type ModelMapper } from "../../modelmap.js";
import { validateRequest } from "../../ir/index.js";
import type {
  Block,
  ImageBlock,
  Request,
  Response,
  ToolChoice,
} from "../../ir/index.js";

export interface NonstreamOptions {
  readonly modelMapper?: ModelMapper;
}

export function decodeRequest(
  wire: JsonObject,
  options: NonstreamOptions = {},
): ConversionResult<Request> {
  const maxTokens = whole(wire.max_tokens, "max_tokens");
  if (maxTokens <= 0n) fail("max_tokens must be positive");
  const losses: Loss[] = [];
  if (wire.metadata !== undefined)
    losses.push(loss("metadata", "metadata", "unmapped-field"));
  const system: { readonly type: "text"; readonly text: string }[] = [];
  if (typeof wire.system === "string") {
    system.push({ type: "text", text: wire.system });
  } else if (wire.system !== undefined && wire.system !== null) {
    const blocks = array(wire.system, "system");
    for (let index = 0; index < blocks.length; index += 1) {
      const block = object(blocks[index], `system[${index}]`);
      if (block.type !== "text") fail(`system[${index}].type must be text`);
      system.push({
        type: "text",
        text: string(block.text, `system[${index}].text`),
      });
      if (block.cache_control !== undefined)
        losses.push(
          loss(
            `system[${index}].cache_control`,
            "cache_control",
            "unmapped-field",
          ),
        );
    }
  }
  const messages = array(wire.messages, "messages").map((entry, index) => {
    const message = object(entry, `messages[${index}]`);
    const role = string(message.role, `messages[${index}].role`);
    if (role !== "user" && role !== "assistant")
      fail(`messages[${index}]: unknown role ${role}`);
    const normalizedRole: "user" | "assistant" = role;
    const content = decodeContent(
      message.content,
      `messages[${index}].content`,
      losses,
    );
    if (content.length === 0) content.push({ type: "text", text: "" });
    return { role: normalizedRole, content };
  });
  const tools = optionalArray(wire.tools, "tools").map((entry, index) => {
    const tool = object(entry, `tools[${index}]`);
    return {
      name: string(tool.name, `tools[${index}].name`),
      ...(tool.description === undefined
        ? {}
        : {
            description: string(
              tool.description,
              `tools[${index}].description`,
            ),
          }),
      input_schema: object(tool.input_schema, `tools[${index}].input_schema`),
    };
  });
  const toolChoice = decodeToolChoice(wire.tool_choice, losses);
  const params = {
    max_tokens: maxTokens,
    ...(wire.temperature === undefined
      ? {}
      : { temperature: number(wire.temperature, "temperature") }),
    ...(wire.top_p === undefined ? {} : { top_p: number(wire.top_p, "top_p") }),
    ...(wire.stop_sequences === undefined
      ? {}
      : {
          stop_sequences: strings(wire.stop_sequences, "stop_sequences"),
        }),
  };
  const request: Request = {
    model: mapModel(options.modelMapper, string(wire.model, "model")),
    ...(system.length === 0 ? {} : { system }),
    messages,
    ...(tools.length === 0 ? {} : { tools }),
    ...(toolChoice === undefined ? {} : { tool_choice: toolChoice }),
    params,
  };
  validateRequest(request);
  return { value: request, losses };
}

export function decodeResponse(
  wire: JsonObject,
  options: NonstreamOptions = {},
): ConversionResult<Response> {
  const losses: Loss[] = [];
  const content = decodeContent(wire.content, "content", losses);
  const nativeStop = string(wire.stop_reason, "stop_reason");
  let stopReason: Response["stop_reason"];
  if (
    nativeStop === "end_turn" ||
    nativeStop === "max_tokens" ||
    nativeStop === "stop_sequence" ||
    nativeStop === "tool_use" ||
    nativeStop === "refusal"
  )
    stopReason = nativeStop;
  else {
    stopReason = "other";
    losses.push(loss("stop_reason", "stop_reason", "unmapped-value"));
  }
  const usage = object(wire.usage, "usage");
  return {
    value: {
      id: string(wire.id, "id"),
      model: mapModel(options.modelMapper, string(wire.model, "model")),
      content,
      stop_reason: stopReason,
      ...(stopReason === "stop_sequence" && wire.stop_sequence !== undefined
        ? {
            stop_sequence: string(wire.stop_sequence, "stop_sequence"),
          }
        : {}),
      usage: {
        input_tokens: whole(usage.input_tokens, "usage.input_tokens"),
        output_tokens: whole(usage.output_tokens, "usage.output_tokens"),
      },
    },
    losses,
  };
}

export function encodeRequest(
  request: Request,
  options: NonstreamOptions = {},
): ConversionResult<JsonObject> {
  validateRequest(request);
  const losses: Loss[] = [];
  if (
    request.metadata !== undefined &&
    Object.keys(request.metadata).length > 0
  )
    losses.push(loss("metadata", "metadata", "unmapped-field"));
  const maxTokens = request.params?.max_tokens ?? 4096n;
  if (request.params?.max_tokens === undefined)
    losses.push(loss("params.max_tokens", "max_tokens", "degraded"));
  const shorthand =
    (request.system === undefined || request.system.length === 0) &&
    request.messages.length === 1 &&
    request.messages[0]!.content.length === 1 &&
    request.messages[0]!.content[0]!.type === "text";
  const messages = request.messages.map((message, index) => ({
    role: message.role,
    content: shorthand
      ? (message.content[0] as { readonly text: string }).text
      : encodeContent(message.content, `messages[${index}].content`, losses),
  }));
  const tools =
    request.tools?.map((tool) => ({
      name: tool.name,
      ...(tool.description === undefined
        ? {}
        : { description: tool.description }),
      input_schema: tool.input_schema,
    })) ?? [];
  const toolChoice = encodeToolChoice(request.tool_choice);
  const params = request.params;
  return {
    value: {
      model: mapModel(options.modelMapper, request.model),
      max_tokens: integer(maxTokens),
      ...(request.system === undefined || request.system.length === 0
        ? {}
        : {
            system: request.system.map((block) => ({
              type: "text",
              text: block.text,
            })),
          }),
      messages,
      ...(params?.temperature === undefined
        ? {}
        : { temperature: fromValue(params.temperature) }),
      ...(params?.top_p === undefined
        ? {}
        : { top_p: fromValue(params.top_p) }),
      ...(params?.stop_sequences === undefined ||
      params.stop_sequences.length === 0
        ? {}
        : { stop_sequences: [...params.stop_sequences] }),
      ...(tools.length === 0 ? {} : { tools }),
      ...(toolChoice === undefined ? {} : { tool_choice: toolChoice }),
    },
    losses,
  };
}

export function encodeResponse(
  response: Response,
  options: NonstreamOptions = {},
): ConversionResult<JsonObject> {
  if (
    response.stop_reason !== "end_turn" &&
    response.stop_reason !== "max_tokens" &&
    response.stop_reason !== "stop_sequence" &&
    response.stop_reason !== "tool_use" &&
    response.stop_reason !== "refusal"
  )
    fail(`stop reason ${response.stop_reason} is unsupported`);
  const losses: Loss[] = [];
  return {
    value: {
      id: response.id,
      type: "message",
      role: "assistant",
      model: mapModel(options.modelMapper, response.model),
      content: encodeContent(response.content, "content", losses),
      stop_reason: response.stop_reason,
      ...(response.stop_reason === "stop_sequence" &&
      response.stop_sequence !== undefined
        ? { stop_sequence: response.stop_sequence }
        : {}),
      usage: {
        input_tokens: integer(response.usage.input_tokens),
        output_tokens: integer(response.usage.output_tokens),
      },
    },
    losses,
  };
}

function decodeContent(
  value: JsonValue | undefined,
  path: string,
  losses: Loss[],
): Block[] {
  if (typeof value === "string") return [{ type: "text", text: value }];
  const blocks: Block[] = [];
  for (let index = 0; index < array(value, path).length; index += 1) {
    const block = object(array(value, path)[index], `${path}[${index}]`);
    const type = string(block.type, `${path}[${index}].type`);
    if (type === "text") {
      blocks.push({
        type: "text",
        text: string(block.text, `${path}[${index}].text`),
      });
    } else if (type === "image") {
      const source = object(block.source, `${path}[${index}].source`);
      if (source.type === "base64")
        blocks.push({
          type: "image",
          media_type: string(
            source.media_type,
            `${path}[${index}].source.media_type`,
          ),
          data: string(source.data, `${path}[${index}].source.data`),
        });
      else if (source.type === "url")
        blocks.push({
          type: "image",
          url: string(source.url, `${path}[${index}].source.url`),
        });
      else {
        losses.push(loss(`${path}[${index}]`, "type", "unsupported-semantic"));
        continue;
      }
    } else if (type === "tool_use") {
      const inputPath = `${path}[${index}].input`;
      const input = object(block.input, inputPath);
      const inputText =
        block.inputText === undefined
          ? (sourceTextOf(input) ?? jsonText(stringifyJson(input)))
          : toolInputText(block.inputText, `${path}[${index}].inputText`);
      blocks.push({
        type: "tool_use",
        id: string(block.id, `${path}[${index}].id`),
        name: string(block.name, `${path}[${index}].name`),
        input: inputText,
      });
    } else if (type === "tool_result") {
      blocks.push({
        type: "tool_result",
        tool_use_id: string(block.tool_use_id, `${path}[${index}].tool_use_id`),
        content: decodeContent(
          block.content,
          `${path}[${index}].content`,
          losses,
        ),
        ...(block.is_error === undefined
          ? {}
          : {
              is_error: boolean(block.is_error, `${path}[${index}].is_error`),
            }),
      });
    } else {
      losses.push(loss(`${path}[${index}]`, "type", "unsupported-semantic"));
      continue;
    }
    if (block.cache_control !== undefined)
      losses.push(
        loss(
          `${path}[${index}].cache_control`,
          "cache_control",
          "unmapped-field",
        ),
      );
  }
  return blocks;
}

function encodeContent(
  blocks: readonly Block[],
  path: string,
  losses: Loss[],
): JsonValue[] {
  const output: JsonValue[] = [];
  for (let index = 0; index < blocks.length; index += 1) {
    const block = blocks[index]!;
    if (block.type === "text") output.push({ type: "text", text: block.text });
    else if (block.type === "image") {
      const image = encodeImage(block);
      if (image === undefined)
        losses.push(loss(`${path}[${index}]`, "image", "unsupported-semantic"));
      else output.push(image);
    } else if (block.type === "tool_use") {
      const value = parseJson(block.input);
      const inputPath = `${path}[${index}].input`;
      output.push({
        type: "tool_use",
        id: block.id,
        name: block.name,
        input: withSourceText(object(value, inputPath), block.input),
      });
    } else {
      const content: JsonValue[] = [];
      for (let child = 0; child < block.content.length; child += 1) {
        const nested = block.content[child]!;
        if (nested.type === "text")
          content.push({ type: "text", text: nested.text });
        else if (nested.type === "image") {
          const image = encodeImage(nested);
          if (image === undefined)
            losses.push(
              loss(
                `${path}[${index}].content[${child}]`,
                "image",
                "unsupported-semantic",
              ),
            );
          else content.push(image);
        } else
          losses.push(
            loss(
              `${path}[${index}].content[${child}]`,
              "content",
              "unsupported-semantic",
            ),
          );
      }
      output.push({
        type: "tool_result",
        tool_use_id: block.tool_use_id,
        content,
        ...(block.is_error === undefined ? {} : { is_error: block.is_error }),
      });
    }
  }
  return output;
}

function encodeImage(image: ImageBlock): JsonObject | undefined {
  if (image.data !== undefined && image.url !== undefined) return undefined;
  if (image.data !== undefined && image.media_type !== undefined)
    return {
      type: "image",
      source: {
        type: "base64",
        media_type: image.media_type,
        data: image.data,
      },
    };
  if (image.url !== undefined && image.media_type === undefined)
    return {
      type: "image",
      source: { type: "url", url: image.url },
    };
  return undefined;
}

function decodeToolChoice(
  value: JsonValue | undefined,
  losses: Loss[],
): ToolChoice | undefined {
  if (value === undefined || value === null) return undefined;
  const choice = object(value, "tool_choice");
  if (choice.disable_parallel_tool_use !== undefined)
    losses.push(
      loss(
        "tool_choice.disable_parallel_tool_use",
        "disable_parallel_tool_use",
        "unmapped-field",
      ),
    );
  const type = string(choice.type, "tool_choice.type");
  if (type === "auto" || type === "any" || type === "none")
    return { mode: type };
  if (type === "tool")
    return {
      mode: "tool",
      name: string(choice.name, "tool_choice.name"),
    };
  losses.push(loss("tool_choice", "tool_choice", "unsupported-semantic"));
  return undefined;
}

function encodeToolChoice(
  value: ToolChoice | undefined,
): JsonObject | undefined {
  if (value === undefined) return undefined;
  if (value.mode === "tool") {
    if (value.name === "") fail("tool_choice.name must not be empty");
    return { type: "tool", name: value.name };
  }
  return { type: value.mode };
}

function loss(path: string, field: string, reason: LossReason): Loss {
  return { path, field, reason };
}
function optionalArray(
  value: JsonValue | undefined,
  name: string,
): readonly JsonValue[] {
  return value === undefined || value === null ? [] : array(value, name);
}
function strings(
  value: JsonValue | undefined,
  name: string,
): readonly string[] {
  return array(value, name).map((entry) => string(entry, name));
}
function toolInputText(value: JsonValue, name: string): JsonText {
  const source = string(value, name);
  try {
    object(parseJson(source), name);
  } catch {
    fail(`${name} must encode an object`);
  }
  return jsonText(source);
}
function array(
  value: JsonValue | undefined,
  name: string,
): readonly JsonValue[] {
  if (!isJsonArray(value)) fail(`${name} must be an array`);
  return value;
}
function object(value: JsonValue | undefined, name: string): JsonObject {
  if (
    value === undefined ||
    value === null ||
    typeof value !== "object" ||
    isJsonArray(value) ||
    isJsonNumber(value)
  )
    fail(`${name} must be an object`);
  return value;
}
function string(value: JsonValue | undefined, name: string): string {
  if (typeof value !== "string") fail(`${name} must be a string`);
  return value;
}
function boolean(value: JsonValue | undefined, name: string): boolean {
  if (typeof value !== "boolean") fail(`${name} must be a boolean`);
  return value;
}
function number(value: JsonValue | undefined, name: string): number {
  if (!isJsonNumber(value)) fail(`${name} must be a number`);
  const result = Number(value.token);
  if (!Number.isFinite(result)) fail(`${name} must be finite`);
  return result;
}
function whole(value: JsonValue | undefined, name: string): bigint {
  if (!isJsonNumber(value) || !value.isInteger)
    fail(`${name} must be an integer`);
  return BigInt(value.token);
}
function fail(message: string): never {
  throw new OxaError("type-violation", `anthropic: ${message}`);
}
